Reorganize Concourse Pipeline: re-assign responsibilities of packaging job

cjcjameson opened the pull request:
Until now, the packaging job has served three purposes:
    - consolidating the fact that a binary passed all ICW testing (or tests at similar levels)
    - being a gatekeeper for long-running & resource-intensive jobs, often those on Pulse
    - producing a wide variety of packages

The packaging job happened to serve the role as gatekeeper because the pulse jobs run against packaged installers. This is still often true, but we want to emphasize instead how shorter running jobs come earlier in the pipeline whereas longer running jobs come later.

This also moves the responsibility, "consolidate the verification that various ICW and similar tests pass", onto the long-running-tests gatekeeper, as it fits better with that job's purpose.

Note that most long running tests use the packages, but a few use the binaries. This is a good thing and affirms our decision to move the responsibility. This change also highlights that many artifacts do not depend on long-running tests!

Because of this change, the installer files in S3 will not necessarily correspond to binaries which passed ICW. If this is concerning, we could implement a promotion mechanism for these artifacts.

Pipeline demonstrating this change: http://gpdb.ci.pivotalci.info/teams/dev/pipelines/pipeline_reorganization_packaging_reorder_1797

cjcjameson commented:
Please refer to #1795 for a different change to the pipeline, in the same pursuit and spirit. These two pull requests are not immediately mergeable, but the ideas behind each could be implemented together. Let us know if you have thoughts on that.

adam8157 commented:
It's OK with me.

ashwinstar commented:
In general I like this restructuring. But do not agree to few points
mentioned below and hence would like to propose some modifications to
what's proposed.

On Fri, Feb 17, 2017 at 4:48 PM, C.J. Jameson <notifications@github.com>
wrote:

> Until now, the packaging job has served three purposes:
> - consolidating the fact that a binary passed all ICW testing (or tests at
> similar levels)
> - being a gatekeeper for long-running & resource-intensive jobs, often
> those on Pulse
> - producing a wide variety of packages
>
> The packaging job happened to serve the role as gatekeeper because the
> pulse jobs run against packaged installers.
>
Agree.

> This is still often true, but we want to emphasize instead how shorter
> running jobs come earlier in the pipeline whereas longer running jobs come
> later.
>
I do not think we should have such hierachical model. Now its only single
set of tests in ICW and rest. Earlier ICW (aka old installcheck-good) was
created for and hence was treated as subset of sanity tests and should pass
to trigger fat tests later. It's not the case now and we wish to move away
from that concept. ICW passing shouldn't be any more treated as gatekeeper
to validate further tests running on pulse. Actually long running tests are
really more important to be triggered and run in parallel to ICW to
cut-done waiting time to see full validation of the commit. There is no
reason to believe if ICW failed long running tests are going to fail as
there is no co-realtion if no duplicate tests exists and its not subset.

> This also moves the responsibility, "consolidate the verification that
> various ICW and similar tests pass", onto the long-running-tests
> gatekeeper, as it fits better with that job's purpose.
>
Hence I would like to propose shouldn't have "long-runnng-test-gatekeeper",
as its anyways misnomer is current context as well. Let other projects flow
through as soon as packages are ready for them to run.

> Note that most long running tests use the packages, but a few use the
> binaries. This is a good thing and affirms our decision to move the
> responsibility. This change also highlights that many artifacts do not
> depend on long-running tests!
>
> Because of this change, the installer files in S3 will not necessarily
> correspond to binaries which passed ICW. If this is concerning, we could
> implement a promotion mechanism for these artifacts.
>
I do not feel that's concerning. But yes would be definitely helpful for
clarity to have job at end which provides visibility of everything passing
on pipeline for a commit.

Thank You !

Chibin commented:
> Hence I would like to propose shouldn't have "long-runnng-test-gatekeeper",
> as its anyways misnomer is current context as well. Let other projects flow
> through as soon as packages are ready for them to run.

I really like this idea. I think we should run all the pulse tests immediately as soon as the packaging completes as long as we can ensure that ICW does not catch anything that may cause the "long-running" tests to fail. 

> I do not think we should have such hierachical model. Now its only single
> set of tests in ICW and rest. Earlier ICW (aka old installcheck-good) was
> created for and hence was treated as subset of sanity tests and should pass
> to trigger fat tests later. It's not the case now and we wish to move away
> from that concept. ICW passing shouldn't be any more treated as gatekeeper
> to validate further tests running on pulse. Actually long running tests are
> really more important to be triggered and run in parallel to ICW to
> cut-done waiting time to see full validation of the commit. There is no
> reason to believe if ICW failed long running tests are going to fail as
> there is no co-realtion if no duplicate tests exists and its not subset.

There was a time when we actually ran all the packaged tests along side installcheck-good and bugbuster and would prefer to go back to that.

My only concern is that we don't have enough resources to keep running all the pulse instances. We will need some kind of mechanism -- we probably have something like this already... -- to ensure that we only run a couple at a time as they will get triggered way quicker. 

Could the "gatekeepers" just be a way to visualize and easily pause specific types of tests, but not really tied to a team. For example, if I want to pause all catalog tests or backup and restore tests, it would be convenient if I could just pause a specific "gatekeeper". Doing it through a script is also doable.

khuddlefish commented:
+1 For moving the packaging job to be concurrent with ICW.

cjcjameson commented:
@ashwinstar great feedback, thank you! I'll rework the PR as follows:
- no gatekeeper for long-running tests; actually try to get those kicked off sooner rather than later
- add a end-of-pipeline job where it's easy to see what has passed "everything", because when looking in the resource-view, you might see all green but not know if that's everything.

@pivotal-mike @tom-meyer @schubert for thoughts on resource availability and tradeoffs

For long-running tests for which resources are a concern, or they simply take too long, we should continue to move those until the nightly-trigger until people have improved the tests.

@Chibin for the further thoughts on gatekeepers, sounds like that's more like #1795 -- feel free to comment there.

pivotal-mike commented:
@cjcjameson I don't see much there to worry about w.r.t. resources.  Moving the long running tests up earlier shouldn't make much if any difference at that level.

cjcjameson commented:
Thank you for your comments! We've addressed them as follows, in order of commits:

- undo the addition of the `long-running-gatekeeper`; that attempt got squashed away and the resulting commit is hopefully more straightforward. The key lines are here: https://github.com/greenplum-db/gpdb/pull/1797/commits/59d7aeb1b8abfc0376e67c7bb77b87c534a04b25#diff-23089528236de2be35566d49df7011caL504
- change all jobs that take longer than 60 minutes to only be triggered nightly.
    @khuddlefish you are affected with 4 jobs now being nightly-trigger only
    @adam8157 you are affected with 2 jobs now being nightly-trigger only
    @Chibin you are affected only for MU_gpexpand
- Provide jobs which summarize success at each stage of the pipeline. This adds noise to the pipeline but could be valuable for the reasons @ashwinstar mentioned at the end of his feedback above. Unfortunately, we need 4 different "summarize" jobs in order to get the look & feel for the pipeline to be correct. The commit HEAD^ does `fly set-pipeline` cleanly, if you want to see the before/after of how much noise is added.

https://gpdb.ci.pivotalci.info/teams/dev/pipelines/pipeline_reorganization_packaging_reorder_1797 is still the place to see what the pipeline would look like!

khuddlefish commented on concourse/pipelines/pipeline.yml:
```diff
@@ -693,7 +721,9 @@ jobs:
 - name: MU_backup-restore
   plan:
-  - aggregate: *pulse_trigger_resource
+  - get: nightly-trigger
```
We would like this to continue being triggered on every commit

khuddlefish commented on concourse/pipelines/pipeline.yml:
```diff
@@ -795,7 +846,9 @@ jobs:
 - name: MU_gptransfer-5x-to-5x
   plan:
-  - aggregate: *pulse_trigger_resource
+  - get: nightly-trigger
```
We would like this to run on every commit too

cjcjameson commented:
@khuddlefish updated that commit to have those two run on every commit (they typically take 70 minutes, so not too far away from being under an hour).

Indeed, the success summarizers seem redundant -- I'm not super committed to them. The challenge is that as a release manager, if you want to know that a particular code sha has gone green on "everything", looking at the resource only gets you part of the answer. If folks have ideas, we're all ears

ashwinstar commented:
regression_tests_gphdfs_centos - runs natively on concourse so I don't think should tie any job to nightly trigger which runs natively, the more we can run every commit its better. Nightly provides very delayed feedback on master/head for Stable branches its fine.

Summary jobs: Sorry for increased work. Didn't think will complicate and add so much noise to pipeline and seems tricky to maintain as well in long run. So, even though I proposed something like that earlier looking at it now definitely feel should loosen that out. We can see without it what need is felt to be added later interatively.

cjcjameson commented:
@ashwinstar no worries man, super-easy to just roll the summarize commit off the top.

I've switched regression_tests_gphdfs_centos back to triggering on every commit.

Updated view, much simpler: https://gpdb.ci.pivotalci.info/teams/dev/pipelines/pipeline_reorganization_packaging_reorder_1797

Chibin commented:
> Updated view, much simpler: https://gpdb.ci.pivotalci.info/teams/dev/pipelines/pipeline_reorganization_packaging_reorder_1797

Liking the new view!

cjcjameson commented:
Yeah -- heads up for @pivotal-mike @tom-meyer @schubert: when there's a product desire for the pipeline to push artifacts to a consolidated delivery place, you will be reintroducing the same underlying idea of "summarizing success". And with the pipeline as it is right now (a two-layered wide fan-out), you will need some of the multi-layer summarizers to get Concourse to render the pipeline in the stages we're used to. How's best for me to keep the knowledge around, or should we just recreate it when we need it? Right now, it's just in my local git reflog.

Would it be helpful for release managers to have a "summarize-success layer" that they can drop onto special release pipelines, if used?
